x/subscription/keeper/grpc_query_estimated_rewards.go

```diff
@@ -0,0 +1,128 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	epochstoragetypes "github.com/lavanet/lava/v2/x/epochstorage/types"
+	rewardstypes "github.com/lavanet/lava/v2/x/rewards/types"
+	"github.com/lavanet/lava/v2/x/subscription/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (k Keeper) EstimatedRewards(goCtx context.Context, req *types.QueryEstimatedRewardsRequest) (*types.QueryEstimatedRewardsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	res := types.QueryEstimatedRewardsResponse{}
+
+	_, specContribut := k.specKeeper.GetContributorReward(ctx, req.ChainId)
+
+	storage := k.epochstorageKeeper.GetAllStakeEntriesCurrentForChainId(ctx, req.ChainId)
+
+	totalStake := math.ZeroInt()
+	var entry epochstoragetypes.StakeEntry
+	found := false
+	for _, e := range storage {
+		totalStake = totalStake.Add(e.EffectiveStake())
+		if e.Address == req.Provider {
+			found = true
+			entry = e
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("provider not found in stake entries for chain ID: %s", req.ChainId)
+	}
+
+	// delegation of the provider
+	delegationAmount := entry.Stake
+	delegatorPart := sdk.NewDecFromInt(delegationAmount.Amount).QuoInt(totalStake).MulInt64(int64(entry.DelegateCommission)).QuoInt64(100)
+	if req.Delegator != "" || req.Amount != "" {
+		if req.Amount != "" {
+			var err error
+			delegationAmount, err = sdk.ParseCoinNormalized(req.Amount)
+			if err != nil {
+				return nil, err
+			}
+			entry.DelegateTotal = entry.DelegateTotal.Add(delegationAmount)
+		} else if req.Delegator != "" {
+			d, found := k.dualstakingKeeper.GetDelegation(ctx, req.Delegator, req.Provider, req.ChainId, uint64(ctx.BlockHeight()))
+			if !found {
+				return nil, fmt.Errorf("could not find delegator")
+			}
+			delegationAmount = d.Amount
+		}
+
+		delegatorPart = sdk.NewDecFromInt(delegationAmount.Amount).QuoInt(totalStake).MulInt64(int64(100 - entry.DelegateCommission)).QuoInt64(100)
+		// calculate existing delegator rewards
+		if entry.DelegateLimit.Amount.LT(entry.DelegateTotal.Amount) {
+			delegatorPart = delegatorPart.MulInt(entry.DelegateLimit.Amount).QuoInt(entry.DelegateTotal.Amount)
+		}
+	}
+	delegatorPart = delegatorPart.Mul(sdk.OneDec().Sub(specContribut))
```

Comment:
in the case of provider, this is not the full provider part, he is also getting the delegators (or limit) multiplied by comission

Reply:
done